lib/publish-rsr.js

```diff
@@ -0,0 +1,96 @@
+import * as Sentry from '@sentry/node'
+import timers from 'node:timers/promises'
+import pRetry from 'p-retry'
+
+const ONE_HOUR = 60 * 60 * 1000
+
+const withPgClient = fn => async ({ createPgClient, ...args }) => {
+  const pgClient = await createPgClient()
+  try {
+    return await fn({ pgClient, ...args })
+  } finally {
+    await pgClient.end()
+  }
+}
+
+export const postEvaluate = withPgClient(async ({ pgClient, round, gitCommit }) => {
+  await pgClient.query(`
+    INSERT INTO publish_rsr_rounds
+    (round_index, spark_evaluate_version, measurement_commitments, round_details, providers)
+    VALUES
+    ($1, $2, $3, $4, $5)
+  `, [
+    round.index,
+    gitCommit,
+    round.cids,
+    round.details,
+    round.measurements.reduce((acc, m) => {
+      acc[m.minerId] = acc[m.minerId] || { successful: 0, total: 0 }
+      acc[m.minerId].total++
+      if (m.retrievalResult === 'OK') {
+        acc[m.minerId].successful++
+      }
+      return acc
+    }, {})
+  ])
+})
+
+const publishRsr = withPgClient(async ({ pgClient, storachaClient, rsrContract }) => {
+  const { rows } = await pgClient.query(`
+    SELECT *
+    FROM publish_rsr_rounds
+    WHERE evaluated_at <= now()::date AND evaluated_at > now()::date - interval '1 day'
+    ORDER BY round_index
+  `)
+
+  const providers = new Map()
+  for (const row of rows) {
+    for (const [minerId, { successful, total }] of Object.entries(row.providers)) {
+      if (!providers.has(minerId)) {
+        providers.set(minerId, { successful: 0, total: 0 })
+      }
+      providers.get(minerId).successful += successful
+      providers.get(minerId).total += total
+    }
+  }
+
+  const directoryCid = await pRetry(() => storachaClient.uploadDirectory([
+    new File([JSON.stringify({
+      date: rows[0].evaluated_at.toISOString().split('T')[0],
+      meta: {
+        rounds: Object.fromEntries(rows.map(row => [
+          row.round_index,
+          {
+            sparkEvaluateVersion: row.spark_evaluate_version,
+            measurementCommitments: row.measurement_commitments,
```

Comment:
Should we use IPLD links here?

```
measurementCommitments: row.measurement_commitments.map(cid => {['/']: cid})
```

Or perhaps serialise the data using https://www.npmjs.com/package/@ipld/dag-json, I think that should handle CID links.

This is not strictly required; it depends on how much we want to be compatible with the rest of the IPLD tooling.

Reply:
https://space-meridian.slack.com/archives/C07CPEGCTFT/p1729776103560939 (private slack)

Reply:
It's IPLD serialized as dag-json now